### PR TITLE
Creation of webservice

### PR DESCRIPTION
**Pull request by tabeckers**
**What issue does this PR address?**
This PR addresses the first coding sprint with regard to the creation of our Webservice. 

**Does this PR introduce a breaking change?**
No breaking changes will be added by this PR. A new project has been added and appended to the FunderMaps solution. A few entities have been appended to the FunderMaps.Core project without modifying anything else.

**Other information**:
A package reference so GeoJSON was added to the FunderMaps.Core project.

### CODE REVIEW

File: FunderMaps.Webservice.Core/Abstraction/Services/IProductService.cs [hunk 1]
```diff
@@ -9,12 +9,12 @@ namespace FunderMaps.Webservice.Abstractions.Services
     /// </summary>
     public interface IProductService
     {
-        Task<BuildingComplete> GetBuildingCompleteByIdAsync(string id);
+        Task<BuildingComplete> GetBuildingCompleteByIdAsync(string id, uint pageNumber, uint pageCount);
```
Comment (yorickdewid) [inline FunderMaps.Webservice.Core/Abstraction/Services/IProductService.cs:12]: We have an interface for this, see INagivation in the core.

File: FunderMaps.Webservice.Core/FunderMaps.Webservice.Core.csproj [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+  </PropertyGroup>
+
+</Project>
```
Comment (yorickdewid) [inline FunderMaps.Webservice.Core/FunderMaps.Webservice.Core.csproj:7]: For all future projects just target .NET Core LTS. Its unlikely we'll need to port some back to .NET Framework

File: FunderMaps.Webservice.Core/Models/BuildingProducts/BuildingBase.cs [hunk 3]
```diff
@@ -1,9 +1,9 @@
-﻿namespace FunderMaps.Webservice.Models.Building
+﻿namespace FunderMaps.Webservice.Core.Models.Building
```
Comment (yorickdewid) [inline FunderMaps.Webservice.Core/Models/BuildingProducts/BuildingBase.cs:1]: Will all these things move to the core eventually? Is that the idea?

File: src/FunderMaps.Core/Exceptions/ProductNotFoundException.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
+﻿using System;
+
+namespace FunderMaps.Core.Exceptions
+{
+    /// <summary>
+    /// Exception indicating we can't find or parse a FunderMaps product.
+    /// </summary>
+    public sealed class ProductNotFoundException : Exception
```
Comment (yorickdewid) [inline src/FunderMaps.Core/Exceptions/ProductNotFoundException.cs:8]: Inherit from core exception

File: src/FunderMaps.Core/Extensions/GuidExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    public static class GuidExtensions
+    {
+        /// <summary>
+        /// Throws an <see cref="ArgumentNullException"/> if the <paramref name="self"/>
+        /// Guid is either null or empty.
+        /// </summary>
+        /// <param name="self"><see cref="Guid"/></param>
+        public static void ThrowIfNullOrEmpty(this Guid self)
+        {
+            if (self == null || self == Guid.Empty) { throw new ArgumentNullException(nameof(self)); }
```
Comment (yorickdewid) [inline src/FunderMaps.Core/Extensions/GuidExtensions.cs:17]: Unfold line

File: src/FunderMaps.Core/FunderMaps.Core.csproj [hunk 6]
```diff
@@ -20,6 +20,7 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="GeoJSON.Net" Version="1.2.19" />
```
Comment (yorickdewid) [inline src/FunderMaps.Core/FunderMaps.Core.csproj:23]: Why do we need this?

File: src/FunderMaps.Core/Types/Years.cs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        /// <summary>
+        /// Constructor based on two years.
+        /// </summary>
+        /// <param name="yearFrom"><see cref="Year"/></param>
+        /// <param name="yearTo"><see cref="Year"/></param>
+        public Years(Year yearFrom, Year yearTo)
+        {
+            YearFrom = yearFrom ?? throw new ArgumentNullException(nameof(yearFrom));
+            YearTo = yearTo ?? throw new ArgumentNullException(nameof(yearTo));
+            if (YearFrom.YearValue >= yearTo.YearValue) { throw new ArgumentOutOfRangeException(nameof(yearFrom)); }
```
Comment (yorickdewid) [inline src/FunderMaps.Core/Types/Years.cs:24]: Newline

File: src/FunderMaps.Core/Types/Years.cs [hunk 8]
```diff
@@ -0,0 +1,57 @@
+﻿using System;
+
+namespace FunderMaps.Core.Types
+{
+    /// <summary>
+    /// Represents a collection of years.
+    /// </summary>
+    public sealed class Years
```
Comment (yorickdewid) [inline src/FunderMaps.Core/Types/Years.cs:8]: Why do we need this? We're not doing `DateTime` again

File: src/FunderMaps.Webservice/Abstractions/Services/IMappingService.cs [hunk 9]
```diff
@@ -0,0 +1,28 @@
+﻿using FunderMaps.Core.Types.Products;
+using FunderMaps.Webservice.ResponseModels;
+using System.Collections.Generic;
+
+namespace FunderMaps.Webservice.Abstractions.Services
+{
+    /// <summary>
+    /// Handles our DTO mapping.
+    /// </summary>
+    public interface IMappingService
```
Comment (yorickdewid) [inline src/FunderMaps.Webservice/Abstractions/Services/IMappingService.cs:10]: AutoMapper

File: src/FunderMaps.Webservice/Controllers/AnalysisController.cs [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+            IProductResultService productResultService)
+        {
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _productResultService = productResultService ?? throw new ArgumentNullException(nameof(productResultService));
+        }
+
+        /// <summary>
+        /// Gets one or more <see cref="ResponseModelBase"/> items in a wrapper. 
+        /// Specify one of the following:
+        /// - Query string <paramref name="q"/>
```
Comment (yorickdewid) [inline src/FunderMaps.Webservice/Controllers/AnalysisController.cs:37]: Use <list><item> when making summations.

File: src/FunderMaps.Webservice/Controllers/AnalysisController.cs [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+        /// <param name="product">The requested product type</param>
+        /// <param name="q">Query string</param>
+        /// <param name="id">Internal building id</param>
+        /// <param name="bagid">BAG id</param>
+        /// <param name="fullFence">Get all buildings in the geofence</param>
+        /// <param name="page">Page to display</param>
+        /// <param name="limit">Items per page</param>
+        /// <returns><see cref="ResponseWrapper{TResponseModel}"/></returns>
+        [HttpGet("get")]
+        public async Task<IActionResult> GetProductAsync(string product, string q, string id, string bagid, bool fullFence, uint? page, uint? limit)
```
Comment (yorickdewid) [inline src/FunderMaps.Webservice/Controllers/AnalysisController.cs:55]: Use an input DTO with the attribute [FromQuery] in the controller

File: src/FunderMaps.Webservice/Controllers/AnalysisController.cs [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+        {
+            // Check for product.
+            if (string.IsNullOrEmpty(product))
```
Comment (yorickdewid) [inline src/FunderMaps.Webservice/Controllers/AnalysisController.cs:58]: Validate in DTO.

File: src/FunderMaps.Webservice/Controllers/AnalysisController.cs [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+            {
+                return BadRequest(Problem("Please specify a product"));
+            }
+
+            // Check for invalid combinations.
+            if ((fullFence && ArgumentUtility.NotNullCount(q, id, bagid) > 0) ||
```
Comment (yorickdewid) [inline src/FunderMaps.Webservice/Controllers/AnalysisController.cs:64]: Use a range check validator in de DTO

File: src/FunderMaps.Webservice/Controllers/AnalysisController.cs [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+                (ArgumentUtility.NotNullCount(q, id, bagid) > 1) ||
+                (!fullFence && ArgumentUtility.NotNullCount(q, id, bagid) == 0))
+            {
+                return Problem("Please select one of the following options: id, bagid, query or fullfence");
+            }
+
+            try
+            {
+                var analysisProduct = ProductTypeMapper.MapAnalysisFromString(product);
```
Comment (yorickdewid) [inline src/FunderMaps.Webservice/Controllers/AnalysisController.cs:73]: Move abstraction and logic further into the core/local managers

File: src/FunderMaps.Webservice/Controllers/StatisticsController.cs [hunk 15]
```diff
@@ -1,16 +1,105 @@
...
+using FunderMaps.Webservice.Mapping;
+using FunderMaps.Webservice.ResponseModels;
 using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Threading.Tasks;
 
 namespace FunderMaps.Webservice.Controllers
 {
-    public class StatisticsController : Controller
```
Comment (yorickdewid) [inline src/FunderMaps.Webservice/Controllers/StatisticsController.cs:9]: See comments on other controllers

File: src/FunderMaps.Webservice/Services/DebugProductService.cs [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace FunderMaps.Webservice.Services
+{
+    /// <summary>
+    /// Used for testing purposes.
+    /// TODO Remove.
+    /// </summary>
+    public sealed class DebugProductService : IProductService
```
Comment (yorickdewid) [inline src/FunderMaps.Webservice/Services/DebugProductService.cs:15]: What is this class?

File: src/FunderMaps.Webservice/Utility/ArgumentUtility.cs [hunk 17]
```diff
@@ -0,0 +1,17 @@
+﻿using System.Linq;
+
+namespace FunderMaps.Webservice.Utility
+{
+    /// <summary>
+    /// Contains utility functions regarding collections of arguments.
+    /// </summary>
+    internal static class ArgumentUtility
```
Comment (yorickdewid) [inline src/FunderMaps.Webservice/Utility/ArgumentUtility.cs:8]: If this is usefull, move it into the core

**Comment by tabeckers** [comment]
All comments have been processed.

**Pull request merged**